### PR TITLE
add IMU_GYRO_RATEMAX to limit gyro control publication rate

### PR DESCRIPTION
**Pull request by dagar**
We'll need something like this before v1.10 to keep the cpu usage under control for certain configurations on older boards.

### CODE REVIEW

**Comment by dagar** [comment]
Cases that come to mind.
 - UAVCAN actuators (a bit of a cpu hog at the moment)
 - stm32f4 boards and VTOL (running both FW + MC + VTOL controllers)

File: src/lib/drivers/gyroscope/PX4Gyroscope.cpp [hunk 1]
```diff
@@ -123,12 +123,26 @@ PX4Gyroscope::update(hrt_abstime timestamp, float x, float y, float z)
...
 	sensor_gyro_control_s &control = _sensor_gyro_control_pub.get();
-	control.timestamp_sample = timestamp;
-	val_filtered.copyTo(control.xyz);
-	control.timestamp = hrt_absolute_time();
-	_sensor_gyro_control_pub.update();	// publish
+
+	if (_param_imu_gyro_rate_max.get() > 0) {
+		const uint64_t interval = 1000000.0f / _param_imu_gyro_rate_max.get();
+
+		if (hrt_elapsed_time(&control.timestamp_sample) >= interval) {
```
Comment (jlecoeur) [inline src/lib/drivers/gyroscope/PX4Gyroscope.cpp]:
```suggestion
		if (hrt_elapsed_time(&control.timestamp_sample) < interval) {
```

File: src/modules/sensors/sensor_params.c [hunk 2]
```diff
@@ -222,12 +221,25 @@ PARAM_DEFINE_INT32(SENS_EN_THERMAL, -1);
 */
 PARAM_DEFINE_FLOAT(IMU_GYRO_CUTOFF, 30.0f);
 
+/**
+* Gyro control data maximum publication rate
+*
+* This is the maximum rate the gyro control data (sensor_gyro_control) will be allowed to publish at.
+* Set to 0 to disable to publish and publish at the native sensor sample rate.
```
Comment (jlecoeur) [inline src/modules/sensors/sensor_params.c]:
```suggestion
* Set to 0 to disable and publish at the native sensor sample rate.
```

File: src/lib/drivers/gyroscope/PX4Gyroscope.cpp [hunk 3]
```diff
@@ -123,12 +123,26 @@ PX4Gyroscope::update(hrt_abstime timestamp, float x, float y, float z)
...
 	// publish control data (filtered gyro) immediately
+	bool publish_control = true;
 	sensor_gyro_control_s &control = _sensor_gyro_control_pub.get();
-	control.timestamp_sample = timestamp;
-	val_filtered.copyTo(control.xyz);
-	control.timestamp = hrt_absolute_time();
-	_sensor_gyro_control_pub.update();	// publish
+
+	if (_param_imu_gyro_rate_max.get() > 0) {
+		const uint64_t interval = 1000000.0f / _param_imu_gyro_rate_max.get();
```
Comment (jlecoeur) [inline src/lib/drivers/gyroscope/PX4Gyroscope.cpp]:
```suggestion
		const uint64_t interval = 1e6f / _param_imu_gyro_rate_max.get();
```

File: src/lib/drivers/gyroscope/PX4Gyroscope.cpp (hunk 1 above)
Comment (dagar) [reply src/lib/drivers/gyroscope/PX4Gyroscope.cpp]: Ha, apparently I shouldn't try to hack these things together on the go. Thanks @jlecoeur

**Comment by dagar** [comment]
> Sounds good, tested with the requested changes on Pixhawk 4.
> 
> It is a bit disappointing to realize that the publication rate does not match the value of the parameter (it is equal the inverse of a multiple of the native publication rate), but I guess that this is the best behaviour.

I think we'll be able to somewhat improve this soon with updated drivers that provide the raw data at higher rates.

For example if we aim for 2 kHz (500 us interval) default on the common Invensense sensors now that are roughly updating at 1 kHz we'll be able to cleanly handle a 400 Hz (2500 us interval) inner loop for PWM setups.

**Pull request merged**